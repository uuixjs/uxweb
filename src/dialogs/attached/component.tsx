import { staticTokenRule, styleVariant, styled } from "@uuixjs/uuixweb-lib";

// TODO: Rename the enum
export enum BalloonDirection {
  Top = "top",
  Bottom = "bottom",
  Left = "left",
  Right = "right",
  TopLeft = "top-left",
  TopRight = "top-right",
  BottomRight = "bottom-right",
  BottomLeft = "bottom-left",
  TopCenter = "top-center",
  BottomCenter = "bottom-center",
  LeftCenter = "left-center",
  RightCenter = "right-center",
}

export interface AttachedProps {
  /** Direction the child will go. */
  direction?:
    | BalloonDirection
    | "top"
    | "bottom"
    | "left"
    | "right"
    | "top-left"
    | "top-right"
    | "bottom-right"
    | "bottom-left"
    | "top-center"
    | "bottom-center"
    | "left-center"
    | "right-center"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** Offsets the child position horizontally; string should include a CSS unit such as '10px' */
  offsetX?: string;
  /** Offsets the child position vertically; string should include a CSS unit such as '10px' */
  offsetY?: string;
}

const positionedMargin = "8px";

/**
 * Applies CSS Absolute positioning to the child to appear just outside of
 * the nearest parent positiooning context, such as position: relative.
 */
export const Attached = styled.div.attrs<AttachedProps>(
  ({ direction = BalloonDirection.Bottom }) => ({
    direction,
  }),
)<AttachedProps>`
  position: absolute;
  z-index: ${staticTokenRule("z-index-balloon")};

  ${styleVariant("direction", {
    [BalloonDirection.Top]: {
      top: "auto",
      bottom: "100%",
      left: 0,
      "margin-bottom": positionedMargin,
    },
    [BalloonDirection.Bottom]: {
      top: "100%",
      left: 0,
      "margin-top": positionedMargin,
    },
    [BalloonDirection.Left]: {
      top: 0,
      right: "100%",
      "margin-right": positionedMargin,
    },
    [BalloonDirection.Right]: {
      top: 0,
      left: "100%",
      "margin-left": positionedMargin,
    },
    [BalloonDirection.TopLeft]: {
      top: "auto",
      bottom: "100%",
      "margin-bottom": positionedMargin,
      right: "auto",
      left: 0,
    },
    [BalloonDirection.TopRight]: {
      top: "auto",
      bottom: "100%",
      "margin-bottom": positionedMargin,
      right: 0,
      left: "auto",
    },
    [BalloonDirection.BottomRight]: {
      top: "100%",
      "margin-top": positionedMargin,
      right: 0,
      left: "auto",
    },
    [BalloonDirection.BottomLeft]: {
      top: "100%",
      "margin-top": positionedMargin,
      right: "auto",
      left: 0,
    },
    [BalloonDirection.TopCenter]: {
      top: "auto",
      bottom: "100%",
      "margin-bottom": positionedMargin,
      left: "50%",
      transform: "translateX(-50%)",
    },
    [BalloonDirection.BottomCenter]: {
      top: "100%",
      "margin-top": positionedMargin,
      left: "50%",
      transform: "translateX(-50%)",
    },
    [BalloonDirection.LeftCenter]: {
      right: "100%",
      "margin-right": positionedMargin,
      top: "50%",
      transform: "translateY(-50%)",
    },
    [BalloonDirection.RightCenter]: {
      left: "100%",
      "margin-left": positionedMargin,
      top: "50%",
      transform: "translateY(-50%)",
    },
  })}

  ${(props) => {
    let marginLeft: string | undefined;
    let marginRight: string | undefined;
    let marginTop: string | undefined;
    let marginBottom: string | undefined;

    if (props.offsetX) {
      if (
        props.direction === BalloonDirection.Left ||
        props.direction === BalloonDirection.TopLeft ||
        props.direction === BalloonDirection.BottomLeft ||
        props.direction === BalloonDirection.LeftCenter ||
        props.direction === BalloonDirection.BottomRight ||
        props.direction === BalloonDirection.TopRight
      ) {
        marginRight = props.offsetX;
      } else {
        marginLeft = props.offsetX;
      }
    }

    if (props.offsetY) {
      if (
        props.direction === BalloonDirection.Top ||
        props.direction === BalloonDirection.TopLeft ||
        props.direction === BalloonDirection.TopRight ||
        props.direction === BalloonDirection.TopCenter
      ) {
        marginBottom = props.offsetY;
      } else {
        marginTop = props.offsetY;
      }
    }

    return {
      marginLeft,
      marginRight,
      marginTop,
      marginBottom,
    };
  }}
`;

Attached.displayName = "Attached";
