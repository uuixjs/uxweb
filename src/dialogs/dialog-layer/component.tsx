import * as PopperJS from "@popperjs/core";

import {
  FC,
  MutableRefObject,
  KeyboardEvent as ReactKeyboardEvent,
  MouseEvent as ReactMouseEvent,
  useCallback,
  useEffect,
  useState,
} from "react";
import { PopperProps, usePopper } from "react-popper";
import { Transition, TransitionProps, TransitionType } from "../../transition";

import ReactModal from "react-modal";
import { ScReactModal } from "../utils/sc-react-modal";
import { cn } from "@uuixjs/uuixweb-lib";
import { useConfigureReactModal } from "../utils/use-configure-react-modal";
import { useDelayedUnmount } from "../utils/use-delayed-unmount";

export interface DialogLayerProps {
  /** Show the dialog when true, hide the dialog when false */
  show: boolean;
  /** Ref to the element the dialog should be positioned next to, usually the `<Button>` which opens the dialog when clicked. */
  anchorRef: PopperProps<{}>["referenceElement"] | null;
  /** Accepts any valid Popper.js options. See https://popper.js.org/docs */
  options?: Partial<PopperJS.Options>;
  /** Allows you to obtain a Ref to the underlying Popper.js instance */
  popperRef?: MutableRefObject<ReturnType<typeof usePopper> | null>;

  /** Callback when user clicks outside dialog or presses escape key */
  onRequestClose:
    | ((event: ReactMouseEvent | ReactKeyboardEvent | MouseEvent) => void)
    | null;
  /** Callback when dialog enter transition has completed. */
  onEnterComplete?: () => void;
  /** Callback when dialog exit transition has completed. */
  onExitComplete?: () => void;

  transitionType?: TransitionProps["type"];
  transitionDuration?: TransitionProps["duration"];
  disableCloseOnClickOut?: boolean;
  disableCloseOnEsc?: boolean;
  disableFocusAfterRender?: boolean;
  disableReturnFocusAfterClose?: boolean;
  /**
   * By default, the application root receives the “aria-hidden” attribute while
   * a dialog is open. Use this to prevent adding the “aria-hidden” attribute
   * while the dialog is open.
   */
  disableAriaHideApp?: boolean;

  portalClassName?: string;

  role?: string;
  ["aria-label"]?: string;
  ["aria-labelledby"]?: string;
  ["aria-describedby"]?: string;
  ["aria-modal"]?: boolean | "false" | "true";
}

export const DialogLayer: FC<DialogLayerProps> = (props) => {
  useConfigureReactModal();

  const { isMounted, onExitComplete } = useDelayedUnmount(
    props.show,
    props.onExitComplete,
  );

  if (!isMounted) {
    return null;
  }

  return (
    <PortaledPopperWithTransition {...props} onExitComplete={onExitComplete} />
  );
};

// Needs to be its own component so that the usePopper hook runs effect cleanup when it is unmounted
const PortaledPopperWithTransition: FC<DialogLayerProps> = ({
  popperRef,
  anchorRef,
  onRequestClose,
  disableCloseOnClickOut,
  disableAriaHideApp,
  ...props
}) => {
  const [dialogRef, setDialogRef] = useState<HTMLDivElement | null>(null);

  const popperInstance = usePopper(anchorRef, dialogRef, {
    placement: "bottom-end",
    ...props.options,
  });

  useEffect(() => {
    if (popperRef) {
      popperRef.current = popperInstance;
    }
  }, [popperInstance, popperRef]);

  const onClickOut = useCallback(
    (e: MouseEvent) => {
      if (disableCloseOnClickOut) {
        return;
      }
      if (
        e.target &&
        anchorRef &&
        anchorRef instanceof Element &&
        anchorRef.contains(e.target as HTMLElement)
      ) {
        return;
      }
      if (onRequestClose) {
        onRequestClose(e);
      }
    },
    [anchorRef, onRequestClose, disableCloseOnClickOut],
  );

  useOnClickOutside(dialogRef, onClickOut);

  return (
    <ScReactModal
      // Aria props:
      role={props.role}
      contentLabel={props["aria-label"]}
      aria={{
        labelledby: props["aria-labelledby"],
        describedby: props["aria-describedby"],
      }}
      ariaHideApp={!disableAriaHideApp}
      // Style props:
      style={dialogLayerModalCss}
      portalClassName={cn(props.portalClassName, "tw-dialog-layer")}
      // Behavior props:
      isOpen // always set to true because the parent will handle un-mount when not visible; ReactModal does not remove the portal when not visible
      onRequestClose={onRequestClose || undefined}
      shouldCloseOnEsc={!props.disableCloseOnEsc}
      shouldCloseOnOverlayClick={!disableCloseOnClickOut}
      shouldFocusAfterRender={!props.disableFocusAfterRender}
      shouldReturnFocusAfterClose={!props.disableReturnFocusAfterClose}
    >
      <div
        ref={setDialogRef}
        style={popperInstance.styles.popper}
        {...popperInstance.attributes.popper}
      >
        <Transition
          type={props.transitionType || TransitionType.Fade}
          duration={props.transitionDuration}
          show={props.show}
          transitionInitialMount
          onExitComplete={props.onExitComplete}
          onEnterComplete={props.onEnterComplete}
        >
          {props.children}
        </Transition>
      </div>
    </ScReactModal>
  );
};

const dialogLayerModalCss: ReactModal.Styles = {
  overlay: {
    position: "fixed",
    top: 0,
    left: 0,
    /**
     * Set width/height to 1px because even though we don't want the overlay to cover anything,
     * this is how react-modal determines if child contents are tabbable or not and having some
     * height/width set helps react-modal understand the content is visible:
     *
     * See https://github.com/reactjs/react-modal/blob/v3.11.2/src/helpers/tabbable.js#L15-L28
     */
    width: 1,
    height: 1,
  },
  content: {
    /** BEGIN TEMPORARY SHIM, TO BE REMOVED BY https://jira.xarth.tv/browse/COREUI-3376  */
    color: "var(--color-text-base)",
    /** END TEMPORARY SHIM */
  },
};

function useOnClickOutside(
  ref: HTMLElement | null,
  handler: (e: MouseEvent) => void,
) {
  useEffect(() => {
    const listener = (event: MouseEvent) => {
      if (!ref || ref.contains(event.target as Node)) {
        return;
      }

      handler(event);
    };

    document.addEventListener("mousedown", listener);

    return () => {
      document.removeEventListener("mousedown", listener);
    };
  }, [ref, handler]);
}
