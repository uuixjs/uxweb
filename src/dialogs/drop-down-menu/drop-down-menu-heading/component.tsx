import { Color, FontSize, Layout } from "../../../layout";
import { CoreText, TextTransform } from "../../../core-text";

import { FC } from "react";
import { getAriaProps } from "@uuixjs/uuixweb-lib";
import { withOverlayContext } from "../../../overlay-region/context";

export interface DropDownMenuHeadingProps {
  children?: string;
  overlay?: boolean;
}

export const DropDownMenuHeadingComponent: FC<DropDownMenuHeadingProps> = (
  props,
) => {
  return (
    <Layout padding={{ x: 0.5 }} margin={{ y: 0.5 }} {...getAriaProps(props)}>
      <CoreText
        color={props.overlay ? Color.Overlay : Color.Alt2}
        fontSize={FontSize.Size6}
        transform={TextTransform.Uppercase}
        bold
      >
        {props.children}
      </CoreText>
    </Layout>
  );
};

DropDownMenuHeadingComponent.displayName = "DropDownMenuHeadingComponent";

export const DropDownMenuHeading = withOverlayContext(
  DropDownMenuHeadingComponent,
);
