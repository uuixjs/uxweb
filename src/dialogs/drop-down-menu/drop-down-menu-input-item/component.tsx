import { AlignItems, Display, InjectLayout, Layout } from "../../../layout";
import {
  DropDownMenuItemFigure,
  DropDownMenuItemFigureProps,
} from "../drop-down-menu-item-figure";
import { newUUIDv4, styled } from "lib";

import { CheckBox } from "../../../form/checkbox";
import { FC } from "react";
import { FormHint } from "../../../form/hint";
import { InputControlProps } from "../../../form/form";
import { Radio } from "../../../form/radio";
import { Toggle } from "../../../form/toggle";

// It seems that even though these are enum members, there is still a
// possibility that they can shadow variables in some cases. See:
// https://github.com/typescript-eslint/typescript-eslint/issues/325
export enum DropDownMenuInputItemType {
  CheckBox = "checkbox", // eslint-disable-line @typescript-eslint/no-shadow
  Radio = "radio", // eslint-disable-line @typescript-eslint/no-shadow
  Toggle = "toggle", // eslint-disable-line @typescript-eslint/no-shadow
}

export interface DropDownMenuInputItemProps extends InputControlProps {
  label: string;
  type: DropDownMenuInputItemType | "checkbox" | "radio" | "toggle"; // TODO: https://jira.xarth.tv/browse/COREUI-3554;
  hint?: string;
  figure?: DropDownMenuItemFigureProps;
}

const ScDropDownMenuInputItemLabel = styled.label`
  cursor: pointer;
`;

const ScDropDownMenuInputItemHint = styled.div`
  margin-left: 2.6rem;
`;

export const DropDownMenuInputItem: FC<DropDownMenuInputItemProps> = ({
  label,
  type,
  hint,
  figure,
  ...inputProps
}) => {
  const icon = figure && (
    <Layout
      flexShrink={0}
      padding={{
        left: type === DropDownMenuInputItemType.Toggle ? undefined : 1,
        right: type === DropDownMenuInputItemType.Toggle ? 0.5 : undefined,
      }}
    >
      <DropDownMenuItemFigure {...figure} />
    </Layout>
  );

  const generatedId = newUUIDv4();

  return (
    <Layout padding={0.5}>
      <Layout display={Display.Flex} alignItems={AlignItems.Center}>
        {type === DropDownMenuInputItemType.CheckBox && (
          <Layout flexGrow={1}>
            <CheckBox label={label} {...inputProps} />
          </Layout>
        )}
        {type === DropDownMenuInputItemType.Radio && (
          <Layout flexGrow={1}>
            <Radio label={label} {...inputProps} />
          </Layout>
        )}
        {icon}
        {type === DropDownMenuInputItemType.Toggle && (
          <>
            <InjectLayout flexGrow={1} margin={{ right: 2 }}>
              <ScDropDownMenuInputItemLabel
                className="tw-drop-down-menu-input-item__label"
                htmlFor={inputProps.id || generatedId}
              >
                {label}
              </ScDropDownMenuInputItemLabel>
            </InjectLayout>
            <Toggle {...inputProps} id={inputProps.id || generatedId} />
          </>
        )}
      </Layout>
      {hint && (
        <ScDropDownMenuInputItemHint className="tw-drop-down-menu-input-item__hint">
          <FormHint hint={hint} />
        </ScDropDownMenuInputItemHint>
      )}
    </Layout>
  );
};

DropDownMenuInputItem.displayName = "DropDownMenuInputItem";
