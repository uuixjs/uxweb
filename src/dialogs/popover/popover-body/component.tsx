import { Component, ReactNode } from "react";
import { InjectLayout, Overflow } from "../../../layout";
import { PaddingValue, getAriaProps, getDataProps } from "lib";

export interface PopoverBodyProps {
  /** This ID attribute must be the same for all elements for a given popover in order to set the correct aria attribute values. In other words, a Popover, PopoverHeader and PopoverBody should each receive the same string. */
  popoverId: string;
  children?: ReactNode;
  padding?: PaddingValue;
}

export class PopoverBody extends Component<PopoverBodyProps, {}> {
  public static defaultProps: Partial<PopoverBodyProps> = {
    padding: 1,
  };

  public render() {
    return (
      <InjectLayout
        {...getDataProps(this.props)}
        {...getAriaProps(this.props)}
        overflow={Overflow.Auto}
        padding={this.props.padding}
        flexGrow={1}
      >
        <div id={`${this.props.popoverId}-body`}>{this.props.children}</div>
      </InjectLayout>
    );
  }
}
