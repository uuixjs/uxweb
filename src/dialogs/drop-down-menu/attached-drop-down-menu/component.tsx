import { FC } from "react";
import { Display, Layout } from "../../../layout";
import { Attached, AttachedProps, BalloonDirection } from "../../attached";
import {
  DropDownMenuWrapper,
  DropDownMenuWrapperProps,
} from "../drop-down-menu-wrapper";

export type DropDownMenuDirection =
  | BalloonDirection.TopLeft
  | BalloonDirection.TopRight
  | BalloonDirection.BottomLeft
  | BalloonDirection.BottomRight;

export interface AttachedDropDownMenuProps
  extends DropDownMenuWrapperProps,
    AttachedProps {
  /** Direction the `DropDownMenu` balloon will go. */
  direction?: DropDownMenuDirection;
  /**
   * Show the `DropDownMenu`. The `DropDownMenu`'s behavior is controlled by
   * its parent.
   */
  show?: boolean;
}

/**
 * Legacy implementation of DropDownMenu which uses CSS position: absolute
 *
 * @deprecated Please migrate to the new `<DropDownMenu>` component; see Core UI 19.0.0 release notes and docs for migration steps.
 */
export const AttachedDropDownMenu: FC<AttachedDropDownMenuProps> = ({
  show,
  direction,
  offsetX,
  offsetY,
  ...props
}) => (
  // Display.Hide was the legacy way this worked; keeping for compatibility
  <Layout display={show ? Display.Block : Display.Hide}>
    <Attached direction={direction} offsetX={offsetX} offsetY={offsetY}>
      <DropDownMenuWrapper {...props} />
    </Attached>
  </Layout>
);
