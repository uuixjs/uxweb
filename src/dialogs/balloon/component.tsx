import { DialogLayerProps } from "../dialog-layer";
import { withDialogLayer } from "../dialog-layer/with-dialog-layer";
import { BalloonWrapper, BalloonWrapperProps } from "./balloon-wrapper";

export interface BalloonProps extends BalloonWrapperProps, DialogLayerProps {}

export const Balloon = withDialogLayer(BalloonWrapper);
