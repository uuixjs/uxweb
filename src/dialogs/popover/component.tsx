import { DialogLayerProps } from "../dialog-layer";
import { withDialogLayer } from "../dialog-layer/with-dialog-layer";
import { PopoverWrapper, PopoverWrapperProps } from "./popover-wrapper";

export interface PopoverProps extends PopoverWrapperProps, DialogLayerProps {}

export const Popover = withDialogLayer(PopoverWrapper);
