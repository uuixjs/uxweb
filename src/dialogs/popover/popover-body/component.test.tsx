import { setupShallowTest } from "../../../../tests/helpers";
import { PopoverBody, PopoverBodyProps } from "./component";

function getRequiredProps(): PopoverBodyProps {
  return {
    popoverId: "test-popover-id",
  };
}

function getOptionalProps(): PopoverBodyProps {
  return {
    popoverId: "test-popover-id",
    children: "Test Content",
    padding: 0,
  };
}

const setupShallow = setupShallowTest(PopoverBody, getRequiredProps);

describe("PopoverBody", () => {
  it("renders popover body component with the required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders popover body component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });
});
