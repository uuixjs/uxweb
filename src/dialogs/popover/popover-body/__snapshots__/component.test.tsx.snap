// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverBody renders popover body component with all props 1`] = `
<InjectLayout
  flexGrow={1}
  overflow="auto"
  padding={0}
>
  <div
    id="test-popover-id-body"
  >
    Test Content
  </div>
</InjectLayout>
`;

exports[`PopoverBody renders popover body component with the required props 1`] = `
<InjectLayout
  flexGrow={1}
  overflow="auto"
  padding={1}
>
  <div
    id="test-popover-id-body"
  />
</InjectLayout>
`;
